samples/bluetooth/encrypted_advertising/central/src/central_ead.c

```diff
@@ -0,0 +1,518 @@
+	} else {
+		LOG_INF("len : %u", data->data_len);
+		LOG_INF("type: 0x%02x", data->type);
+		LOG_HEXDUMP_INF(data->data, data->data_len, "data:");
+
+		/* Copy the data out if we are running in a test context */
+		if (received_data != NULL) {
+			size_t offset = *((size_t *)user_data);
+
+			offset += bt_data_serialize(data, &received_data[offset]);
+
+			*((size_t *)user_data) = offset;
+		} else {
+			*((size_t *)user_data) += bt_data_get_len(data, 1);
+		}
+	}
+
+	return true;
+}
+
+static void device_found(const bt_addr_le_t *addr, int8_t rssi, uint8_t type,
+			 struct net_buf_simple *ad)
+{
+	int err;
+	size_t parsed_data_size;
+	char addr_str[BT_ADDR_LE_STR_LEN];
+
+	if (default_conn) {
+		return;
+	}
+
+	bt_addr_le_to_str(addr, addr_str, sizeof(addr_str));
+
+	/* We are only interested in the previously connected device. */
+	if (!bt_addr_le_eq(addr, &peer_addr)) {
+		return;
+	}
+
+	LOG_DBG("Peer found.");
+
+	parsed_data_size = 0;
+	LOG_INF("Received data size: %zu", ad->len);
+	bt_data_parse(ad, data_parse_cb, &parsed_data_size);
+
+	LOG_DBG("All data parsed. (total size: %zu)", parsed_data_size);
+
+	err = bt_le_scan_stop();
+	if (err) {
+		LOG_DBG("Failed to stop scanner (err %d)", err);
+		return;
+	}
+
+	k_poll_signal_raise(&device_found_cb_completed, 0);
+}
+
+static void connect_device_found(const bt_addr_le_t *addr, int8_t rssi, uint8_t type,
+				 struct net_buf_simple *ad)
+{
+	int err;
+	char addr_str[BT_ADDR_LE_STR_LEN];
+
+	if (default_conn) {
+		return;
+	}
+
+	/* Connect only to devices in close range */
+	if (rssi < -70) {
+		return;
+	}
+
+	bt_addr_le_to_str(addr, addr_str, sizeof(addr_str));
+
+	LOG_DBG("Device found: %s (RSSI %d)", addr_str, rssi);
+
+	err = bt_le_scan_stop();
+	if (err) {
+		LOG_DBG("Failed to stop scanner (err %d)", err);
+		return;
+	}
+
+	err = bt_conn_le_create(addr, BT_CONN_LE_CREATE_CONN, BT_LE_CONN_PARAM_DEFAULT,
+				&default_conn);
+	if (err) {
+		LOG_DBG("Failed to connect to %s (err %d)", addr_str, err);
+		return;
+	}
+
+	k_poll_signal_raise(&device_found_cb_completed, 0);
+}
+
+static void start_scan(bool connect)
+{
+	int err;
+
+	k_poll_signal_reset(&conn_signal);
+	k_poll_signal_reset(&device_found_cb_completed);
+
+	err = bt_le_scan_start(BT_LE_SCAN_PASSIVE, connect ? connect_device_found : device_found);
+	if (err) {
+		LOG_DBG("Scanning failed to start (err %d)", err);
+		return;
+	}
+
+	LOG_DBG("Scanning started.");
+
+	if (connect) {
+		LOG_DBG("Waiting for connection");
+		await_signal(&conn_signal);
+	}
+
+	await_signal(&device_found_cb_completed);
+}
+
+static uint8_t gatt_read_cb(struct bt_conn *conn, uint8_t att_err,
+			    struct bt_gatt_read_params *params, const void *data, uint16_t read_len)
+{
+	gatt_read_err = att_err;
+	gatt_read_len = read_len;
+	gatt_read_handle = params->by_uuid.start_handle;
+
+	if (!att_err) {
+		memcpy(gatt_read_res, data, read_len);
+	}
+
+	k_poll_signal_raise(&gatt_read_signal, 0);
```

Comment:
In the case we get an ATT error, should we really raise the signal here? Wouldn't an ATT error make everything afterwards not work?